Changes ui-side behavior in response to websocket job status updates on several lists

mabashian opened the pull request:
##### SUMMARY
Part of #5883 

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - UI

##### ADDITIONAL INFORMATION
It might be easiest for me to outline the expected behavior with these changes in order to illustrate what's changed and pose questions.

1. UI gets a socket message with a `status` _other than_ `pending`.
    - UI takes that message, loops across the available rows attempting to find a row that matches up with the `unified_job_id` provided in the socket message.  If it finds a matching row, we update the status of the row to match the status in the message.  (This same pattern will extend out to other fields (started/finished/etc) when provided by the api).

2. UI gets a socket message with at `status` of `pending`.
    - This is where things get interesting
    - This code treats this message like: This is a new job and you don't have it in your list.
        - Although now that I think about it, maybe this isn't quite true.  Could there be scenarios where we _do_ have this job in the list already but it's status is `new`?  If so, the code as it currently stands would potentially insert an extra job for this row (we don't check for this job in our existing list rows).
    - When we get a message like this, we have to go out and perform a GET in order to determine whether or not this particular user has sufficient privileges to see this job _and_ whether or not it fits in with the current search filters.
        - We only do this if the user is on page 1 of the jobs list
        - We only do this if the user has the default `-finished` or similar `-started` `order_by` filter applied to the list.  If they're sorting by something else we have no idea where this potential new row fits into the list.
        - We only do this if the user hasn't already filtered the list down by `id__in`.
            - There's a corner case here - if the user provided an `id__in` filter that matched a future job id.  This probably isn't a realistic scenario though.
        - These GET requests are time-banded on a 5 second interval.  For example, lets say a new job comes in over the websocket and we make a GET request for it: `/api/v2/unified_jobs?id__in=9000&count_disabled=1`.  Over the next 5 seconds, we hold on to all the id's of new jobs that come in and bundle them up into a single request to be executed 5 seconds after the initial request (id=9000).  That second request would look like `/api/v2/unified_jobs?id=9001,9002,9003`&count_disabled=1.  The cycle continues until there are no more new jobs in the past 5 seconds.
        - If we get rows back, this code adds them to the top of the list (this is why the sort order is important).  Page size is maintained by bumping the corresponding number of rows off the bottom.
        - Once the rows are added, future socket messages with updates to these rows fall into scenario 1 above.

**A special note about the launch modal**: while it's open we will still **update** rows underneath the modal (because this no longer requires making GET requests and re-rendering the entire row).  We don't add any new rows to the list while this modal is open though.  This is because the modal itself (it's element) lives within the row.  If this row were to get bumped out of the list the modal would disappear.  As such, we hold on to new job id's and fire off the corresponding request only _after_ the launch modal is closed.  This behavior is consistent with the way we handle messages while the launch modal is open currently.

### Question(s)

Probably for @wenottingham but I'm curious what others think as well.

1) Adding rows to the top of the list when the sort order is `-finished` (the default sort value) is not exactly "right".  In reality, we're adding rows to the top of the list in the order which we receive the socket messages telling us that a new job has been created.  This is probably OK until the job actually finishes.  Once the jobs finish, they may be out of order based on the `-finished` `order_by` filter.  Is that OK?

Alternatively we could try to do something like:

```
[new jobs ordered by whatever order they came in first (or something else?)]
[new jobs that we learned about from sockets but they've finished running, ordered by -finished (ordering would happen by the UI)]
[api jobs ordered by -finished]
```

mabashian committed:
Changes how the jobs list reacts to socket messages.  We now only make targeted GET requests for new rows.  We use the available information in the socket message to update the relevant row (if visible in the list).

wenottingham commented:
> we have to go out and perform a GET in order to determine whether or not this particular user has sufficient privileges to see this job

Shouldn't we be fixing this on the sending side?

ryanpetrello commented:
> Shouldn't we be fixing this on the sending side?

@wenottingham unlike "give me event data about Job XYZ", the "tell me about job status changes" websocket group isn't _currently_ scoped in any way - it's literally just "tell me the IDs of jobs that encounter a status transition".

So if we wanted to change this (**especially** if we wanted to make it include _additional_ data _about_ the job in question), we'd probably have to have some sort of per-user channel, or do RBAC checks for *every* status change for *every* subscriber.  However, even if we changed this, we'd *still* probably have to do a `GET` request if our end goal is to *only* show list views based on currently applied filters.

One of the things we're experimenting with here is the idea that doing a list view and limiting it to `id__in=X,Y,Z` is **really** fast and gives us all the context we need - *hopefully* with fairly minimal API and UI changes compared to supporting an entirely new set of websocket interactions.

chrismeyersfsu commented:
![image](https://user-images.githubusercontent.com/722880/74770215-ce019a80-5259-11ea-8fe6-aa49614393c7.png)
:+1:

AlanCoding commented:
> Could there be scenarios where we do have this job in the list already but it's status is new?

No, we got rid of that. In this flow, you POST to JT jobs list or global jobs list to create a job. Then you POST to the start endpoint. _This was removed, there is no way to do this anymore._ It might still be the case that some tracebacks will put a job into the "new" state, as a vestigial thing. That means the server had an error, and is not a concern for us here.

AlanCoding commented:
> Adding rows to the top of the list when the sort order is -finished (the default sort value) is not exactly "right". In reality, we're adding rows to the top of the list in the order which we receive the socket messages telling us that a new job has been created.

If you're filtering by `-finished` that leaves ambiguous how jobs with a null finished time are sorted amongst themselves. Probably the default API sorting. By the Django ORM, it's possible to provide a secondary order_by column, so you could sort by `-finished`, and then `-created` (or `-id` could proxy for reverse creation order). That could require some extra work on the filter logic, but I think that's what you're moving towards. I don't necessarily see an inconsistency here.

Anyway, I say it's a safe assumption that "pending" status messages are received in order of creation and that successful/fail/error status messages are received in order of finished time, and discrepancies would be so minuscule that there is no value in worrying about it.

+1 from me after reading your synopsis.

jakemcdermott commented:
implementation and described behavior lgtm :+1: 

It's worth pointing out:

> We only do this if the user has the default -finished or similar -started order_by filter applied to the list. If they're sorting by something else we have no idea where this potential new row fits into the list.

I think this means we don't add new rows to lists with search filters applied?  My gut tells me that removing live updates for these is a reasonable trade-off but I'm only guessing. @wenottingham @ryanpetrello @AlanCoding  / anyone else with a bit more insight to customer usage at scale, does this seem like an acceptable trade-off?

wenottingham commented:
Is the implication that right now if you have some filter applied to the list, new rows will appear that *don't* match the filter?

mabashian commented:
@wenottingham no, that shouldn't happen.  We can apply those same filters when making the GET requests for the new rows.  If they don't match the filters, we won't get them back.  The `order_by` is the one exception.  If it's not `-finished` or `-started`, this code won't try to fetch new rows because we wouldn't know where to put them.

wenottingham commented:
So the options are a) not show them b) do a 'full' list request in that case only?

mabashian commented:
@wenottingham yea, that's my understanding.  Those would be the two options.

mabashian committed:
Stop making rest calls for project sync updates on projects list

mabashian committed:
Stop making rest calls for inventory source sync updates on inventory sources list

rebeccahhh committed:
added unified job template ID to websocket emits

mabashian committed:
Removes logic performing GET requests on `api/v2/templates` whenever a job status update message comes across the websocket.  We now use data exclusively from the websocket to update the UI.

mabashian committed:
Removes GET requests in response to websocket messages on the org job templates list.  List is solely updated based on data from the websocket messages.

rebeccahhh committed:
added finished job timestamp to websocket emit

mabashian committed:
Consume finished timestamp from websocket message when it's available to update the relevant row in the templates list.

rebeccahhh committed:
fixed the spelling of cancelled to be canceled, note to us later -> we need to stick with a single spelling of the word

mabashian committed:
Consume finished timestamp from websocket message and update the relevant job row.  Also adds logic to attempt to re-order the list when the sort order is -finished since we have enough information client-side to do that.

mabashian committed:
Consume finished timestamp on org templates list when available via websocket message

mabashian committed:
Trims down GET requests made on the dashboard in response to websocket messages

mabashian committed:
Fix jshint errors

mabashian committed:
Remove tooltip update when job finishes.  This will be handled later down the line by buildTooltips().

mabashian commented:
OK, here's a rundown of all of the lists that subscribe to the `jobs` websocket channel and listen for status changes:

### Jobs List (/#/jobs)
  - How it used to work:
    - GET request to refresh the entire list was made whenever a message came across the websocket but it was time-banded so that we only made this request once every 5 seconds.
  - How it works now:
    - If the job in question already exists in our list, we use the contents of the websocket to update the row.  When a job _finishes_ and the sort order is `-finished` we attempt to re-sort the list client side with the jobs without a finished timestamp at the top.
    - If we get a new job and the sort order is `-finished` or `-started`, we issue a request to see if the user has access to that job (`/api/v2/unified_jobs?id__in=...`).  This request is time-banded to every 5 seconds.  Note that any search parameters applied are maintained in this request.  If rows come back, we add them to the top of the list and trim this list down to match the page size.  If the sort order is something else we don't try to add rows.
### Projects List (/#/projects)
  - How it used to work:
    - GET request to refresh the entire list was made whenever a project sync for a project in the list _finished_.
  - How it works now:
    - Rows are updated using the contents of the websocket message.  No requests are made after initial load.
### Templates List (/#/templates)
  - How it used to work:
    - GET request to refresh the entire list was made whenever a message came across the websocket but it was time-banded so that we only made this request once every 5 seconds.
  - How it works now:
    - For new jobs (status = pending) we try to find the template it belongs to and we add this new job to the front of the recent_jobs array (popping the last job off if length > 10).  For all other status updates we try to find the template+job and update the status/finished timestamp.  No GET requests needed.
### Dashboard List(s) (/#/home)
  - How it used to work:
    - 4 requests (jobs/templates/dashboard/dashboard graph) were made whenever a message came across the websocket but it was time-banded so that we only made these requests once every 5 seconds.
  - How it works now:
    - The two dashboard related requests are still made (time-banded to every 5 seconds) but with one small caveat: these requests are only triggered by a job **finishing**.  I could be wrong about this but I'm not sure that we need to update the dashboard data until a job actually finishes.
    - The Recently Used Templates list is now updated with information from the websocket message with one notable exception: a job finishes that's tied to a template that is _not_ currently in this list.  When that happens, the UX would be that the template in question should appear at the top of this list.  In order to achieve this, we need to make a GET request to see if the user actually has access to this template (`/api/v2/unified_job_templates?id__in=...`).  If so, we add it to the top of the list and pop the bottom row off to maintain 5.  This request is time-banded to every 5 seconds. **Note that this request is not made while the launch modal is open.  This is because the element that houses the launch modal is inside each row.  If the user is launching a JT and that JT gets popped off the array the launch modal would close which is pretty aggravating.  Instead, we hold the request until the launch modal is close at which point we make any pending requests.**
    - The Recent Job Runs list is now updated with the information from the websocket message with one notable exception: a job finishing.  Since this list shows _only_ completed jobs, we have to make a GET request whenever we detect a finished job to see if a user has access to it (`/api/v2/unified_jobs?id__in=...`).  If they do, we add it the top of the list and bump the last job off.  This request is time-banded to every 5 seconds.
### Inventory Sources List (/#/inventories/:id/inventory_sources)
  - How it used to work:
    - GET request to refresh the entire list was made whenever an inventory source sync for a source in the list _finished_.
  - How it works now:
    - Rows are updated using the contents of the websocket message.  No GET requests are made after initial load.
### Organization Job Templates List (/#/organizations/:id/job_templates)
  - How it used to work:
    - GET request to refresh the entire list was made whenever a message came across the websocket.
  - How it works now:
    - For new jobs (status = pending) we try to find the JT it belongs to and we add this new job to the front of the recent_jobs array (popping the last job off if length > 10).  For all other status updates we try to find the JT+job and update the status/finished timestamp.  No GET requests needed.
### Organization Projects List (/#/organizations/:id/projects)
  - How it used to work:
    - Rows were updated using the contents of the websocket message
  - How it works now:
    - Same, I didn't change this file

ryanpetrello commented:
@mabashian 

<img src="https://media1.tenor.com/images/6a207337ff44e92f8dd167f2d99bca6b/tenor.gif?itemid=4881098" />

rebeccahhh reviewed (approved):
@mabashian is the real MVP

rebeccahhh committed:
Added in check for the unified_job_template_id attribute to be present and populated in the object.
For anyone reading this later, know that AdHocCommands still have unified_job_template and unified_job_template_id fields, they are just nonetypes because they don't get used by the AdHocCommand objects. Which means you have to actually get the object, not just check that it's there, to use it the way I am in this change.

jakemcdermott commented:
@mabashian @rebeccahhh :wave: Still LGTM - Where do we stand with this one? Is this thing ready to merge or does it need more review?

mabashian commented:
@jakemcdermott I believe @unlikelyzero is exploring test automation on this PR today.  Once we get the 👍 there I think we'll be ready to merge.

ryanpetrello commented on awx/main/models/unified_jobs.py:
```diff
@@ -1218,12 +1218,17 @@ def _websocket_emit_status(self, status):
                     status_data['instance_group_name'] = self.instance_group.name
                 else:
                     status_data['instance_group_name'] = None
+            elif status in ['successful', 'failed', 'canceled']:
```
@mabashian is the UI resilient to this field not being present in the case of e.g., `status=error` ?

mabashian replied:
The UI shouldn't collapse when we hit a message like that we just won't be able to update that field.  The logic for the jobs list is [here](https://github.com/ansible/awx/pull/5968/files#diff-02aed5f19734ca0e82f923146ab885dbR123).  Basically, we check to see if `finished` is included in the message and if it is we'll update the row.  On the Jobs list in particular the only "problem" this would cause would be that the client-side sorting we're attempting to do may get out of sync with what a real API response would provide.  This is because we're putting the rows without a `finished` timestamp up at the top of the list.

We do have some logic over on the dashboard that uses this same array for determining whether or not a job has finished (prompting us to go out and fetch new rows).  Should `error` be a part of [this array](https://github.com/ansible/awx/pull/5968/files#diff-5c4a87eec8896be28053f47299364133R111)?  I think it probably should?

mabashian replied:
See https://github.com/ansible/awx/pull/5968/commits/f2955d7bb307f1effd680ed7c0bc730a11af347d

ryanpetrello commented on awx/main/models/unified_jobs.py:
```diff
@@ -1218,12 +1218,17 @@ def _websocket_emit_status(self, status):
             status_data.update(self.websocket_emit_data())
             status_data['group_name'] = 'jobs'
+            if getattr(self, 'unified_job_template_id', None):
+                status_data['unified_job_template_id'] = self.unified_job_template_id
```
👍

ryanpetrello reviewed (approved):
The API changes, and @mabashian's description of what changed in the UI make sense to me, and sound super promising.

Thanks to everyone for the hard work on this one, it's going to be well received by users 🎉.

mabashian committed:
Jobs that error should trigger us to to update the Recent Jobs list along with successful/failed/canceled

rebeccahhh committed:
formatted finished time to match microsecond expected output as is in the API.

rebeccahhh committed:
added a check for the field of finished making sure it's not none, and then breaking the DateTime function

jakemcdermott commented:
@unlikelyzero @mabashian @rebeccahhh :wave: Still LGTM - Where do we stand with this one? Is this thing ready to merge or does it need more review / testing?

unlikelyzero commented:
Final stages of testing. Merge meeting is tomorrow
